### PR TITLE
OCPBUGS-7559: Remove hard requirement for the afterburn from early-running aws-related services

### PR DESCRIPTION
**Pull request by lobziik**
In older versions of OCP (4.1 specifically) boot image does not contain the afterburn service.
This causes instance provisioning issues within
clusters that use bootimages from OCP 4.1 and below.

**- What I did**

This commit adds fallback logic into service scripts that try to communicate with IMDS via curl
if afterburn populated environment variables were not found.

**- How to verify it**
Use bootimage from the OCP 4.1 for machine boot. Machine should come successfully.

For ocp 4.12+
1. Update user-data-secret for using ignition 2.2.0 
- ```oc -n openshift-machine-api get -o json secret worker-user-data | jq '.data.userData | @base64d | fromjson' | jq '.networkd = {} | .passwd = {} | .storage = {} | .systemd = {} | .ignition.version = "2.2.0" | .ignition.security.tls.certificateAuthorities[0].verification = {} | .ignition.config.append = .ignition.config.merge | del(.ignition.config.merge) | .ignition.config.append[0].verification = {}' >v2.2.json```
- ```oc -n openshift-machine-api set data secret/worker-user-data --from-file=userData=v2.2.json```
- ```oc -n openshift-machine-api set data secret/worker-user-data-managed --from-file=userData=v2.2.json```
2. Update one of machinesets to use AMI from OCP 4.1
- ```REGION="$(oc get -o json infrastructure cluster | jq -r .status.platformStatus.aws.region)"```
- ```AMI="$(curl -s https://raw.githubusercontent.com/openshift/installer/release-4.1/data/data/rhcos.json | jq -r ".amis[\"${REGION}\"].hvm")"```
- ```oc -n openshift-machine-api patch machineset MACHINESET_NAME --type json -p "[{\"op\": \"add\", \"path\": \"/spec/template/spec/providerSpec/value/ami\", \"value\": {\"id\": \"${AMI}\"}}]"```
3. Trigger machine creation using edited machineset (scale up, or delete existing machineset-managed machine) 

**- Description for the changelog**

Related prs:
- https://github.com/openshift/machine-config-operator/pull/2988
- https://github.com/openshift/machine-config-operator/pull/3170/

### CODE REVIEW

**Comment by lobziik** [comment]
I checked this manually on OCP 4.15.6 with using AMI from OCP 4.1, systemd services works

**Comment by sinnykumari** [comment]
Thanks @lobziik for making this quick fix. This looks good to me.
/approve

Always good to have another set of eyes for review. Can someone from your team also take a look and add lgtm? 

Also, adding hold for pre-merge testing by qe and add qe_approved label
/hold

**Comment by lobziik** [comment]
@sunzhaohua2 , can you please take a look and do a pre-merge testing for this :)

**Comment by sinnykumari** [comment]
> @sinnykumari Can we make sure that QE is prompted to test this ASAP. It'd be excellent if this could begin the long backport process toward 4.11 and once QE has tested this I'll work to help accelerate that process.

These mult-team intersection bugs are confusing that which qe team is going to verify this issue. MCO QE can test it but possible that we may miss specific sub-tests that respective team have better expertise. But I do agree that sooner the fix gets merged it is better. MCO team can test and verify this specific bug.

/cc @rioliu-rh @sergiordlr

**Comment by lobziik** [comment]
To verify this there are 2 cases needed to be checked:
1. New machines should come with 4.1 bootimage (briefly described in pr description)
2. New machines with enabled IMDSv2 should come with 4.1 bootimage
    - to enable imdsv2 for machines, need slightly configure a machineset: https://docs.openshift.com/container-platform/4.11/machine_management/creating_machinesets/creating-machineset-aws.html#machineset-creating-imds-options_creating-machineset-aws

**Comment by lobziik** [comment]
>How critical do you think it is to ensure that the small set of clusters born in 4.1 have the ability to leverage IMDSv2? 

@sdodson, I don't think that its possible to disable this functionality such cluster without breaking other clusters. We already shipped the API for enabling IMDSv2 through MAPI machinesets, and we need to ensure that any vm is configured in a way when it can boot and join the cluster.

This issue need to be fixed anyway, either by removing afterburn dependency or updating bootimage, otherwise newly created machines will not boot properly for clusters which was born as 4.1.

**Comment by lobziik** [comment]
@sdodson if you mean is just simplifying the code inside the unit and remove token request from there, yeah, we can do it. This will break new machines creations for 4.1-born clusters in case if IMDSv2 was enabled via a machineset. I didn't find any information about IMDSv1 deprecation, this should be ok at least for now.

**Review by cgwalters — APPROVED** [review]
The pros/cons here are:

- By taking this approach, we're still using afterburn (proper Rust code) for current OCP versions instead of hand-rolled bash scripts
- But inherently, we're also not testing the "fallback" code in CI here, it needs manual testing (that said, I did try setting up an "old-bootimages" job a while ago...)
- Eventually, once we have proper support for updating bootimages, this code can just be deleted

Whereas the plan of "ignore afterburn and always use curl" would mean that we'd need to try to re-work to use afterburn again after we have bootimage updates.

**Comment by sunzhaohua2** [comment]
> @sunzhaohua2 , can you please take a look and do a pre-merge testing for this :)

Seems doesn't work for me, have added comment in [OCPBUGS-7559](https://issues.redhat.com/browse/OCPBUGS-7559?focusedCommentId=21888297&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-21888297)

**Comment by lobziik** [comment]
I simplified this code a bit, due to old RHCOS (till OCP 4.7) does not support imdsv2, and there is no point to obtain tokens in the fallback code.

PTAL:
https://github.com/openshift/machine-config-operator/compare/8f513e4db4340fc1b4f228115dfe0abd9f60704e..c8c299c2657894a10d1289c15e4a81af6fb1716c

**Comment by lobziik** [comment]
Speaking about verification, there are two new bugs popped up during testing this patch:
- https://issues.redhat.com/browse/OCPBUGS-9969 - MCO bug
- https://issues.redhat.com/browse/OCPBUGS-9971 - MAPI IMDSv2 bug

To verify this specific patch, need to ensure that units `aws-kubelet-nodename.service` and `aws-kubelet-providerid.service` finished successfully during machine boot with the OCP 4.1 bootimage. Machine should be created without IMDSv2 enforced. cc @sunzhaohua2

**Comment by sinnykumari** [comment]
removing hold as this has been verified by qe
/hold cancel
/lgtm

**Comment by sdodson** [comment]
The e2e-aws-ovn-upgrade job is failing due to https://issues.redhat.com/browse/TRT-897. I probably can't override this but will try and seek out someone on Slack to override if not.
/override ci/prow/e2e-aws-ovn-upgrade

**Comment by jkyros** [comment]
Agree with Scott, and if this was breaking something, it would have broken before it got to the prometheus issue. 
/override ci/prow/e2e-aws-ovn-upgrade

**Comment by cgwalters** [comment]
Contrary to mythology then, prometheus didn't cause the fire

**Pull request merged**
